_posts/2024-09-05-FEX-2409.md

```diff
@@ -0,0 +1,84 @@
+for if statements and loops. For example, to check if two variables are equal,
+we subtract them and check the zero flag. To check if one variable is less than
+another, we subtract and check the negative flag.  This pattern -- subtracting,
+setting flags, and discarding the actual result -- is so common it has a
+special instruction: **CMP** ("**C**o**MP**are").
+
+If the story ended here, emulation would be easy. Unfortunately, we need to
+talk about the *carry* flag.
+
+After an addition, the *carry* flag indicates if the result overflowed. The
+flag can then be input to another addition, accelerating 128-bit addition.
+
+Subtractions are similar. In hardware, subtractions are additions with an
+operand negated. Like additions, subtractions set the carry flag to accelerate
+128-bit subtractions. Precisely how is the flag defined? There are two
+competing conventions.
+
+The first sets the flag when there is a *borrow*, by mathematical analogy with
+addition. x86 uses this "borrow flag" convention, as it seems more natural.
+
+The second option sets the flag when there is *not* a borrow. Seem backwards?
+Maybe. Adding a (two's complement) negated operand overflows exactly when the
+subtraction does *not* borrow. This "true carry" convention matches actual
+hardware behaviour.  The "borrow" x86 convention requires extra gates to invert
+carry. Arm uses the "true carry" convention to save a few gates.
+
+Which convention should FEX use?
+
+We could store the x86 carry flag in the Arm carry flag.  Unfortunately, that
+requires an extra instruction after each subtraction to invert carry to get the
+borrow flag.
+
+The counter-intuitive alternative is storing the *opposite* of the x86 flag.
+That requires an extra invert after every *addition*, but it eliminates the
+invert after subtraction.
+
+Either we pay after additions or after subtractions. Which should we pick?
+
+While addition is common, using the *flags* from an addition is not. Flags
+are typically used with comparisons, which are subtractions. Therefore, the
+inverted convention will usually win. This month, Alyssa adjusted FEX to invert
+carries, speeding up typical workloads by a few percent.
+
+She didn't stop there. August was filled with optimization. Overall, benchmarks
+are upwards of 10% faster since the last release.
+
+## A Qt change
+
+What about more user visible changes? If you use the FEXConfig tool to
+configure the emulator, you're in for a treat. While it works, this imgui-based
+tool is not known for its aesthetics. Or usability. Or accessibility. In
+between his work optimizing the [redacted] out of FEX's [redacted], Tony
+rewrote FEXConfig as a simple Qt application, improving all three areas in one
+shot. This month release includes both versions of FEXConfig for you to test.
+
+FEXConfig is often used to configure "speedhacks", options that trade emulation
+accuracy for performance. The new FEXConfig makes using those options a bit
+nicer.
```

Comment:
I feel like certain news sites might take this paragraph too literally - not sure we want people to say "the new fexconfig makes speedhacks easier to use". As a precaution I'd suggest:

```suggestion
Besides better look and feel, we've polished first-time setup for logging, library forwarding, and RootFS images. We've also made tweaking various emulation settings a little bit nicer. Users of our Ubuntu PPA can simply update to unlock these improvements without any further action.
```